### PR TITLE
Add Automated Tests. Travis-CI + CodeCov.io.

### PR DESCRIPTION
**Pull request by chris001**
This Pull Request is to satisfy Feature Request #531 .

To see these tests in action, have a look at the Travis-CI and CodeCov.io status badges:
* https://github.com/EspaceNetworks/offlineimap
* Click on the status badges, for detailed test reports.

Function internetConnected() returns True or False

Simple function to quickly check, are we connected to the internet.  Returns true or false. Needed to solve bug where hangs forever on resume from suspend or sleep, due to no longer connected to internet, or IP address change.

Update internetConnected.py

Simple function to quickly check, are we connected to the internet.
Returns true or false.
Needed to solve bug where hangs forever on resume from suspend or sleep, due to no longer connected to internet, or IP address change.

Add requests.

Requests, used by internetConnected().

Add Build Status badges, for Travis and CodeCov.

Add .travis.yml for automated tests.

Add all 4 links at top of README.md

No linebreak between travis status and codecov.

Add linebreak between top 4 links.

New tests folder for new set of tests.

Bullet list.

Add pytest and unittests stuff for codecov.io

Add pytest codecov coverage pytest-cov

Add .coveragerc

"Solves a number of issues collecting coverage."

Add link to fork's travis codecov status badges.

"coverage run tests.py", unittest already running.

python nightly doesn't run on trusty.

python 3.7 fail on trusty. unittest, testrunner.

python 3.6

Matrix of auth mechanisms.

Clean up extra lines in python matrix.

Temp disable osx.

create_conf_file.py

Add create_conf_file.py

Add refresh token.

Set command line option to use generated config

Update .travis.yml with encrypted gmail and outlook login credentials.
Addd tests/.gitignore to skip .pyc files.

Update create_conf_file.py to read in the variables from environment.

Update create_conf_file.py to copy the conf file to second location for pytest.

Update create_conf_file.py to actually copy the conf file for pytest.

Add missing import in create_conf_file.py

Update create_conf_file.py to actually output the variable for sslcacertfile.

Fix typo in filename ca-certificates.crt

Fix typo in oauth2_client_secret

Add Repository IMAP (duplicate of Repository Outlook) this is hardcoded in OLItest test runner.

Update TestRunner.py to correctly import imaplib2 instead of imaplib.
import offlineimap.virtual_imaplib2 as imaplib

Fix type in import.

Disable pytest (test/OLItest/TestRunner.py) because it hardcodes LOGIN authoentication which fails on Gmail. TODO: Fix pytest login and enable pytest again.

Update shebangs to python2 (bin/offlineimap for installed, and ./offlineimap.py for non-installed)

Fix import ConfigParser for python3 has changed library name to configparser.

Fix for python3: Config.set(sect,'maxsyncaccounts',1) TypeError: option values must be strings.

Comment ssl_version to allow dynamic selection of tls1_1 or tls1_2 (which doesn't seem to be available on Travis python3).

Change tls_secure to tls_compat. This is required when you want the app to dynamically select tls1 tls1_1 or tls1_2, so it will run on Travis python3.

Fix typo in tls_level "tls_compat".

Comment "tls_level" setting. Default value is already "tls_compat".

Add support for separate auth_mechanisms for Outlook (IMAP) and Gmail (which always uses XOAUTH2 currently).

Comment CRAM-MD5 and GSSAPI auth_mechanisms, and change them to Dovecot (Outlook doesn't support them) in anticipation of creating Dovecot IMAP inbox acct.

Try exclude "[Gmail]/All Mail" with 'folderfiler' option.

Ativate OSX testing. Remove python3 from allowed failures.

Add OSX ssl ca cert file.

Fix travis python on mac osx.

Workaround for OSX python for Travis.

Update build matrix.

Travis OSX python workarounds.

More config for OSX.

More OSX Travis python.

Troubleshoot Travis Mac OSX python.

More Travis OSX python.

Fix travis settings.

Fix travis.yml no multi-line if statements.

Fix travis.yml

Fix syntax error in travis.yml

Travis OSX python.

Travis OSX python.

Fixes for Travis MacOS python.

Fix Travis Mac OSX python.

Travis OSX python.

More Travis OSX python.

Travis Mac OSX python.

Travis Mac OSX python.

Travis Mac OSX python.

Travis Mac OSX python skip activate myvenv.

Travis Mac OSX python export ssl ca cert file.

Travis Mac OSX python. Matrix.

Travis Mac OSX full set of OSX variations in "include". TODO: no list of each OSX test, instead use a proper matrix.

> This v1.1 template stands in `.github/`.

### This PR

> Add character x `[x]`.

- [] I've read the [DCO](http://www.offlineimap.org/doc/dco.html).
- [] I've read the [Coding Guidelines](http://www.offlineimap.org/doc/CodingGuidelines.html)
- [] The relevant informations about the changes stands in the commit message, not here in the message of the pull request.
- [] Code changes follow the style of the files they change.
- [] Code is tested (provide details).

### References

- Issue #no_space

### CODE REVIEW

**Comment by nicolas33** [comment]
I can't apply as-is. This shows me as the author and it has my signed-off-by line. The patch is mixed up with the v7.0.13 commit, I guess.

Please, rebase on top of v7.0.13 or the current master/next branch. This should fix the issue.

**Comment by chris001** [comment]
OK, it's now rebased on top of master (7.1.5).  It should pass Travis tests.  Green light.  Ready to apply now.

File: README.md [hunk 1]
```diff
@@ -52,6 +64,14 @@ message without internet connection? No problem, the message is still there.
...
 
 
+## Why should I use OfflineIMAP?
+
+* It is **fast**.
+* It is **reliable**.
+* It is **flexible**.
+* It is **safe**.
+
+
```
Comment (nicolas33) [inline README.md]: Why these changes? Please, see a92b116a4f0b70396049deca779783309de8d897

File: README.md [hunk 2]
```diff
@@ -87,9 +107,9 @@ Bugs, issues and contributions can be requested to both the mailing list or the
 ## Requirements & dependencies
 
 * Python v2.7+
-* Python v3.4+ ***[STALLED] (experimental: [see known issues](https://github.com/OfflineIMAP/offlineimap/issues?q=is%3Aissue+is%3Aopen+label%3APy3))***
+* Python v3.4+ ***(experimental: [see known issues](https://github.com/OfflineIMAP/offlineimap/issues?q=is%3Aissue+is%3Aopen+label%3APy3))***
```
Comment (nicolas33) [inline README.md]: It IS stalled. Nobody is working on this anymore. Please, discard this hunk.

File: README.md [hunk 3]
```diff
@@ -87,9 +107,9 @@ Bugs, issues and contributions can be requested to both the mailing list or the
...
 * six (required)
-* imaplib2 >= 2.57 (optional)
+* imaplib2 >= 2.55 (optional)
```
Comment (nicolas33) [inline README.md]: Incorrect hunk.

File: test/.gitignore [hunk 4]
```diff
@@ -1,2 +1,5 @@
 credentials.conf
-tmp_*
\ No newline at end of file
+tmp_*
```
Comment (nicolas33) [inline test/.gitignore]: No more newline?

File: tests/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+*.pyc
+
```
Comment (nicolas33) [inline tests/.gitignore]: Should not be necessary. It's already in the root .gitignore.

File: offlineimap.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 # Startup from single-user installation
 # Copyright (C) 2002 - 2008 John Goerzen
 # <jgoerzen@complete.org>
```
Comment (nicolas33) [inline offlineimap.py]: This commit will be discarded by a rebase on top of the current public next branch. ,-)

File: README.md (hunk 1 above)
Comment (chris001) [reply README.md]: Those "it is flexible it is safe" lines came in by mistake, because, added the new build status icon badges to the 7.1.13 version of `README.md`. Those lines can be removed.

File: README.md (hunk 2 above)
Comment (chris001) [reply README.md]: OK.

File: README.md (hunk 3 above)
Comment (chris001) [reply README.md]: OK

File: test/.gitignore (hunk 4 above)
Comment (chris001) [reply test/.gitignore]: OK. It wasn't intentional!

File: tests/.gitignore (hunk 5 above)
Comment (chris001) [reply tests/.gitignore]: OK.

File: offlineimap.py (hunk 6 above)
Comment (chris001) [reply offlineimap.py]: OK, no problem!

**Comment by chris001** [comment]
> Looks like the tests are failing but I don't care, for now. We can fix them later.

* The reason for the tests failing here (the upstream repo) is, because of security reasons, the encrypted environmental variables (the mail servers login credentials) are ONLY available in the repo where they were added (fork).  This is a security rule enforced by Travis.
* Look in the repo of origin (the fork), scroll down to the Build Status badge (displayed at the top of the `README.md` at the bottom of the source code), you will see, the Travis tests are  passing: https://github.com/EspaceNetworks/offlineimap
* To solve this, you only need to add the encrypted environmental variables to THIS repo and the tests be able to login, and will pass!
* NOTE: the required environmental variables (login credentials) are totally new.  The variables I previously emailed to you are invalid/expired now.  I will have to email you the new ones, or better, do you have a secure instant messenger.

**Comment by nicolas33** [comment]
Ok. I'm on gitter if we can consider this secure (I think so). If you'd really want secured transfer I can send you a public key.

**Comment by nicolas33** [comment]
https://gitter.im/OfflineIMAP/offlineimap

File: README.md (hunk 2 above)
Comment (chris001) [reply README.md]: On second thought... Shouldn't we keep the "[STALLED]" label for `python3`. 
* Because Travis is testing the `python3` environment. And it's passing.  Even though, it's "cheating" by shebang `python2`.  It will usually run on a `python3` machine. As long as it also has `python2` installed also.
* Travis make this `python3` conversion effort MUCH EASIER, because it gives INSTANT FEEDBACK to EVERY CODE CHANGE on ALL OPERATING SYSTEMS, ALL AUTH MECHANISMS, and ALL PYTHON VERSIONS.  
* Converting to `python3` is a must. There's no choice.  Python3 in 2020 or else the app does not live beyond 2020.  
* The changes needed are relatively small.
* Tools such as `2to3`, `futurize`, and `modernize`, make it even easier.
* This conversion will happen. 
* Anyone can come and submit python3 patches, and Travis immediate feedback will show the validity and if the python3 conversion patches don't also break existing `python2`.

File: README.md (hunk 1 above)
Comment (chris001) [reply README.md]: Lines removed.

File: test/.gitignore (hunk 4 above)
Comment (chris001) [reply test/.gitignore]: The original file (on the left here) had no newline because `tmp_*` was the last line in the file. 
The new file adds a newline because it also adds more lines to the file.

File: README.md (hunk 2 above)
Comment (nicolas33) [reply README.md]: We already got 3 attempts to make offlineimap support py3. All failed with critical issues. The required changes are big and introduce subtle bugs because of bytes and UTF-8.

I don't think the conversion will happen, actually. I think the alternative will be imapfw.

**Comment by nicolas33** [comment]
Let me know when you think the PR is ready for a merge. Please, fix all the requested changes and squash the fixes to the correct patches. ,-)

File: offlineimap.py (hunk 6 above)
Comment (chris001) [reply offlineimap.py]: Leaving in this commit, so that the Travis tests on `python3` will run and pass.  As soon as the encrypted login vars are added.  Without having to wait for the current public next branch.

File: tests/.gitignore (hunk 5 above)
Comment (chris001) [reply tests/.gitignore]: I've found reference, which states that, after they tested this scenario (does a root `.gitignore` pass its ignore to subdirectory), and their test results conclude that `.gitignore` files are not inherited by subdir!  https://stackoverflow.com/a/20880582/4339193

**Comment by chris001** [comment]
@nicolas33 
All requested changes fixed (except the `.gitignore` and the newline, which both have good reasons to leave as it is, see link in comment above).  Have squashed the fixes to the correct patches.  The PR is ready to merge.  Only missing piece is the encrypted IMAP environmental variables.  Will look for you on gitter now.

File: tests/.gitignore (hunk 5 above)
Comment (nicolas33) [reply tests/.gitignore]: No, this line is not necessary.

test is not a nested git repo and the author of the comment you link to is really doing weird things with its own configuration.

File: offlineimap.py (hunk 6 above)
Comment (nicolas33) [reply offlineimap.py]: These changes are already in the current 'next' branch. I was expecting you to rebase on the current 'next' branch. :-)
I you do so, these hunks will disappear.

Comment (nicolas33) [reply offlineimap.py]: Anyway, I'll apply.

**Comment by nicolas33** [comment]
Applied to the temporary `trytravis` branch. Thank you!

**Comment by nicolas33** [comment]
Forgot to say I did minor tuning of both the patch and commit message.
